### PR TITLE
KTOR-3873 Introduce BaseApplicationPlugin

### PR DESCRIPTION
**Pull request by e5l**
Fix [KTOR-3873](https://youtrack.jetbrains.com/issue/KTOR-3873)

### CODE REVIEW

File: ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPlugin.kt [hunk 1]
```diff
@@ -34,17 +34,27 @@ public interface Plugin<
...
 @Suppress("AddVarianceModifier")
-public interface ApplicationPlugin<
+public interface BaseApplicationPlugin<
     in TPipeline : Pipeline<*, ApplicationCall>,
     out TConfiguration : Any,
     TPlugin : Any> : Plugin<TPipeline, TConfiguration, TPlugin>
 
+/**
+ * Defines a Plugin that is installed into Application
+ * @param TPipeline is the type of the pipeline this plugin is compatible with
```
Comment (rsinukov) [inline ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPlugin.kt:50]: please update kdoc

File: ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPlugin.kt [hunk 2]
```diff
@@ -34,17 +34,27 @@ public interface Plugin<
...
+ * @param TConfiguration is the configuration object type for this Plugin
+ * @param TPlugin is the instance type of the Plugin object
+ */
+@Suppress("AddVarianceModifier")
+public interface ApplicationPlugin<out TConfiguration : Any> :
```
Comment (rsinukov) [inline ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPlugin.kt:55]: should it be type alias or do we need a real type?

File: ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPlugin.kt [hunk 3]
```diff
@@ -34,17 +34,27 @@ public interface Plugin<
...
 /**
- * Defines a Plugin that is installed into Application
+ * Defines a Plugin that is installed into Application.
  * @param TPipeline is the type of the pipeline this plugin is compatible with
  * @param TConfiguration is the configuration object type for this Plugin
  * @param TPlugin is the instance type of the Plugin object
  */
 @Suppress("AddVarianceModifier")
-public interface ApplicationPlugin<
+public interface BaseApplicationPlugin<
```
Comment (rsinukov) [inline ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPlugin.kt:43]: what about route scoped plugin?

**Review by rsinukov — CHANGES_REQUESTED** [review]

File: ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPlugin.kt (hunk 2 above)
Comment (e5l) [reply ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPlugin.kt:55]: I will check if the type alias appears in the IDE actions and if so we can use it

File: ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPlugin.kt (hunk 1 above)
Comment (e5l) [reply ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPlugin.kt:50]: Sure, thanks

File: ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPlugin.kt (hunk 3 above)
Comment (e5l) [reply ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPlugin.kt:43]: Sure, let me fix this as well

File: ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPlugin.kt (hunk 2 above)
Comment (e5l) [reply ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPlugin.kt:55]: Checked. Typealias are not used when you specify an explicit type. So interface should be used

**Review by rsinukov — APPROVED** [review]
LGTM, but api check is failed/

**Pull request merged**